### ISSUE TITLE
False positive Undefined name for self-referential return annotation in a PEP 695 generic class

### ISSUE THREAD

**Issue by sam-s** [astral-sh/ruff#27379]
### Bug description

ruff reports `Undefined name uniqdict` when a method in a [PEP 695](https://peps.python.org/pep-0695/) generic class uses the enclosing class in its return annotation.

```python
class uniqdict[K, V](dict[K, V]):
    def __ior__(self, other, /) -> uniqdict[K, V]:
        return self
```

### Command used

```shell
ruff check --output-format concise z.py
```

### Ruff output

```python
z.py:3:35: ;;https://docs.astral.sh/ruff/rules/undefined-name Undefined name `uniqdict`
```

### Expected behavior

no diagnostics

### OS / Environment

MSYS_NT-10.0-26100

### Larger real-world example
```python
import collections.abc

class uniqdict[K, V](dict[K, V]):
    """A mapping whose existing keys cannot be assigned different values."""

    def __setitem__(self, key: K, value: V) -> None:
        try:
            if (old := super().__getitem__(key)) != value:
                raise ValueError(self.__class__.__name__, key, old, value)
        except KeyError:
            super().__setitem__(key, value)

    update = collections.abc.MutableMapping.update

    def __ior__(self, other, /) -> uniqdict[K, V]:
        self.update(other)
        return self
```

### Version

ruff 0.15.22

**Comment by ntBre** [comment]
I initially thought this was related to #27369, but Python 3.13 does emit an error here:

```pycon
Python 3.13.13 (main, Apr  7 2026, 18:19:01) [Clang 21.0.0 (clang-2100.0.123.102)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class uniqdict[K, V](dict[K, V]):
...     def __ior__(self, other, /) -> uniqdict[K, V]:
...         return self
...
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    class uniqdict[K, V](dict[K, V]):
        def __ior__(self, other, /) -> uniqdict[K, V]:
            return self
  File "<python-input-0>", line 1, in <generic parameters of uniqdict>
    class uniqdict[K, V](dict[K, V]):
        def __ior__(self, other, /) -> uniqdict[K, V]:
            return self
  File "<python-input-0>", line 2, in uniqdict
    def __ior__(self, other, /) -> uniqdict[K, V]:
                                   ^^^^^^^^
NameError: name 'uniqdict' is not defined
```

This is only valid on Python 3.14 or with `from __future__ import annotations`:

```pycon
Python 3.13.13 (main, Apr  7 2026, 18:19:01) [Clang 21.0.0 (clang-2100.0.123.102)] on darwin
>>> from __future__ import annotations
>>> class uniqdict[K, V](dict[K, V]):
...     def __ior__(self, other, /) -> uniqdict[K, V]:
...         return self
...
>>>
```

```pycon
Python 3.14.4 (main, Apr  7 2026, 13:13:20) [Clang 21.0.0 (clang-2100.0.123.102)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class uniqdict[K, V](dict[K, V]):
...     def __ior__(self, other, /) -> uniqdict[K, V]:
...         return self
...
```

You may need to set your Python version with either `requires-python` in pyproject.toml, or with the Ruff-specific [target-version](https://docs.astral.sh/ruff/settings/#target-version). Ruff also doesn't emit a diagnostic with this configuration: https://play.ruff.rs/24c8c825-4c7f-45bb-bc91-df188d90153f
